### PR TITLE
Refactor & Increase Test Coverage of Entire Codebase

### PR DESCRIPTION
**Pull request by joyguptaa**
Date: 23 March 2025

Developer Name: Joy Gupta

---

## Issue Ticket Number
- https://github.com/Real-Dev-Squad/discord-service/issues/72

## Description

Increased test coverage across the entire codebase. Refactored closely linked functions to use interfaces, improving modularity, scalability, and testability. This change will simplify future maintenance and enhance testing through easier mocking and dependency injection.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Test Coverage

Screenshot 2025-03-23 at 7 18 42 PM

File having 93.8% coverage
Screenshot 2025-03-23 at 7 33 27 PM

### CODE REVIEW

**Comment by joyguptaa** [comment]
@coderabbitai review

File: models/discord_test.go [hunk 1]
```diff
@@ -41,4 +42,35 @@ func TestSessionWrapper(t *testing.T) {
 		}, "should panic when getUerId() is called")
 	})
 
+	t.Run("SessionWrapper should always implement guildMemberNickname() method", func(t *testing.T) {
+		assert.Panics(t, func() {
+			sessionWrapper.GuildMemberNickname("1", "2")
+		}, "should panic when guildMemberNickname() is called")
+	})
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply models/discord_test.go:49]: why do I even store err if it's going to panic anyway?

File: queue/main.go [hunk 2]
```diff
@@ -11,60 +11,90 @@ import (
...
+			ContentType: "text/plain",
+			Body:        message,
+		},
+	); err != nil {
+		logrus.Errorf("Failed to publish message: %v", err)
+		return err
+	}
+	logrus.Info("Message sent successfully")
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main.go:84]: like how? give me an example

File: queue/main_test.go [hunk 3]
```diff
@@ -78,29 +109,39 @@ func TestGetQueueInstance(t *testing.T) {
...
+		}}
+		assert.NoError(t, mockQueue.CreateChannel())
+	})
+	t.Run("QueueWrapper should always implement DeclareQueue() method", func(t *testing.T) {
+		queueWrapper.Channel = &amqp.Channel{}
 		assert.Panics(t, func() {
-			sessionWrapper.declareQueue()
+			queueWrapper.DeclareQueue()
 		})
 	})
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main_test.go:144]: it cannot be placed in the test. Because this would require an actual queue server running. Otherwise, it will panic

File: services/helloService_test.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
 	"testing"
 
 	"github.com/Real-Dev-Squad/discord-service/fixtures"
-	_ "github.com/Real-Dev-Squad/discord-service/tests/helpers"
+	_ "github.com/Real-Dev-Squad/discord-service/tests/setup"
 	"github.com/Real-Dev-Squad/discord-service/utils"
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply services/helloService_test.go:12]: already verified

File: queue/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main.go:84]: I think logrus.Info is fine too.

File: services/main_test.go [hunk 5]
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/Real-Dev-Squad/discord-service/config"
 	"github.com/Real-Dev-Squad/discord-service/dtos"
 	"github.com/Real-Dev-Squad/discord-service/fixtures"
-	_ "github.com/Real-Dev-Squad/discord-service/tests/helpers"
+	_ "github.com/Real-Dev-Squad/discord-service/tests/setup"
 	"github.com/Real-Dev-Squad/discord-service/utils"
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply services/main_test.go:13]: verified it and its working fine.

File: config/constants_test.go [hunk 6]
```diff
@@ -3,7 +3,7 @@ package config
 import (
 	"testing"
 
-	_ "github.com/Real-Dev-Squad/discord-service/tests/helpers"
+	_ "github.com/Real-Dev-Squad/discord-service/tests/setup"
 	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply config/constants_test.go:7]: verified it and its working fine.

File: services/discordBaseService_test.go [hunk 7]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/Real-Dev-Squad/discord-service/dtos"
 	"github.com/Real-Dev-Squad/discord-service/fixtures"
 	service "github.com/Real-Dev-Squad/discord-service/services"
-	_ "github.com/Real-Dev-Squad/discord-service/tests/helpers"
+	_ "github.com/Real-Dev-Squad/discord-service/tests/setup"
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply services/discordBaseService_test.go:13]: verified it and its working fine.

File: controllers/baseHandler_test.go [hunk 8]
```diff
@@ -6,7 +6,7 @@ import (
 	"testing"
 
 	"github.com/Real-Dev-Squad/discord-service/controllers"
-	_ "github.com/Real-Dev-Squad/discord-service/tests/helpers"
+	_ "github.com/Real-Dev-Squad/discord-service/tests/setup"
 	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply controllers/baseHandler_test.go:10]: verified it and its working fine.

File: commands/handlers/main_test.go [hunk 9]
```diff
@@ -43,52 +44,26 @@ func TestMainHandler(t *testing.T) {
...
-	defer func() { CreateSession = originalCreateSession }()
-
-	t.Run("should return error if newNickName is longer than 32 characters", func(t *testing.T) {
-		err := UpdateNickName("userID", "ThisIsAVeryLongNicknameThatExceedsTheLimit")
+	t.Run("should return error if nickname is more than 32 characters", func(t *testing.T) {
+		sessionWrapper := models.SessionWrapper{}
+		err := UpdateNickName("userID", "nicknameWithMoreThan32Characters.", &sessionWrapper)
 		assert.Error(t, err)
 		assert.Equal(t, "Must be 32 or fewer in length.", err.Error())
 	})
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply commands/handlers/main_test.go]: check now

File: tests/main.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type MockSession struct {
+	OpenError                error
+	CommandError             error
+	ApplicationCommandCalled bool
+	CloseCalled              bool
+	GetUserIdCalled          bool
+	GuildMemberNicknameError bool
+}
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply tests/main.go]: check now

File: tests/main.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+	if m.CommandError == nil {
+		m.ApplicationCommandCalled = true
+	}
+	return nil, m.CommandError
+}
+
+func (m *MockSession) GetUerId() string {
+	m.GetUserIdCalled = true
+	return ""
+}
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply tests/main.go]: check now

File: queue/main_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main_test.go:136]:
```
func (q *QueueWrapper) DeclareQueue() error {
	var err error
	q.Queue, err = q.Channel.QueueDeclare(
		config.AppConfig.QUEUE_NAME,
		true,
		false,
		false,
		false,
		amqp.Table{"x-max-priority": 2},
	)
	return err
}
```

From this code its already evident that it will only return error if its there otherwise nil. Do I really need to mock it first and then tests it to be not nil?

File: queue/main_test.go [hunk 12]
```diff
@@ -6,62 +6,93 @@ import (
...
+
+	t.Run("should not panic when PublishMessage() returns error", func(t *testing.T) {
+		mockQueue := &QueueWrapper{Publish: func(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
+			return nil
+		}}
+		assert.NotPanics(t, func() {
+			mockQueue.PublishMessage([]byte("message"))
+		}, "InitQueueConnection should not panic when DeclareQueue is unsuccessful")
+	})
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main_test.go]: check now

File: queue/main_test.go [hunk 13]
```diff
@@ -6,62 +6,93 @@ import (
...
+
+	t.Run("should return error when PublishMessage() returns error", func(t *testing.T) {
+		mockQueue := &QueueWrapper{Publish: func(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
+			return assert.AnError
+		}}
+		assert.Error(t, mockQueue.PublishMessage([]byte("message")))
+	})
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main_test.go]: check now

File: queue/main_test.go [hunk 14]
```diff
@@ -6,62 +6,93 @@ import (
...
+			InitQueueConnection(mockQueue)
+		}, "InitQueueConnection should not panic when CreateChannel is unsuccessful")
+	})
+	t.Run("should not panic when Connection.Channel() returns error", func(t *testing.T) {
+		mockQueue := &MockQueue{ChannelError: errors.New("channel failed")}
 		assert.NotPanics(t, func() {
 			InitQueueConnection(mockQueue)
 		}, "InitQueueConnection should not panic when CreateChannel is unsuccessful")
-
 	})
```
[the remark this answers is not in the document]
Comment (joyguptaa) [reply queue/main_test.go]: done check now

**Review by iamitprakash — CHANGES_REQUESTED** [review]
"No, its just a test case updation or addition" Please follow formate Yes or No

File: commands/handlers/listeningHandler.go [hunk 15]
```diff
@@ -15,5 +16,9 @@ func (s *CommandHandler) listeningHandler() error {
 	} else {
 		nickName = strings.TrimPrefix(strings.TrimSuffix(nickName, utils.NICKNAME_SUFFIX), utils.NICKNAME_PREFIX)
 	}
-	return UpdateNickName(s.discordMessage.UserID, nickName)
+	sessionWrapper, err := models.CreateSession()
```
Comment (iamitprakash) [inline commands/handlers/listeningHandler.go:19]: description talks about only test case, then why we need change into business logic, if needed please attached ticket link

File: commands/handlers/listeningHandler_test.go [hunk 16]
```diff
@@ -58,6 +79,6 @@ func TestListeningHandler(t *testing.T) {
 		handler := &CommandHandler{discordMessage: dataPacket}
 		err := handler.listeningHandler()
 		assert.Error(t, err)
-		assert.Equal(t, "websocket: close 4004: Authentication failed.", err.Error())
+		assert.Equal(t, "HTTP 401 Unauthorized, {\"message\": \"401: Unauthorized\", \"code\": 0}", err.Error())
```
Comment (iamitprakash) [inline commands/handlers/listeningHandler_test.go:82]: what is this code about - "\"code\": 0"

File: commands/handlers/main.go [hunk 17]
```diff
@@ -34,39 +32,14 @@ func MainHandler(dataPacket []byte) func() error {
 	}
 }
 
-type DiscordSession struct {
-	session *discordgo.Session
-}
```
Comment (iamitprakash) [inline commands/handlers/main.go:39]: same here

File: commands/handlers/main_test.go [hunk 18]
```diff
@@ -43,52 +44,21 @@ func TestMainHandler(t *testing.T) {
...
+		mockSess := &tests.MockSession{ForceErrorForGuildMemberNickname: true}
+		err := UpdateNickName("userID", "nickname", mockSess)
 		assert.Error(t, err)
-		assert.Equal(t, "failed to create session", err.Error())
 	})
-	t.Run("should hit GuildMemberNickname if CreateSession succeeds", func(t *testing.T) {
-		CreateSession = mockCreateSession
-		assert.Panics(t, func() { UpdateNickName("userID", "validNickname") })
-
+	t.Run("should not return error if GuildMemberNickname succeeds", func(t *testing.T) {
```
Comment (iamitprakash) [inline commands/handlers/main_test.go:59]: I do see +ve and -ve tests, what about re-try

File: commands/main/register.go
Comment (iamitprakash) [inline commands/main/register.go:1]: why this file got deleted

File: commands/register/register.go [hunk 19]
```diff
@@ -4,14 +4,13 @@ import (
...
+	"github.com/Real-Dev-Squad/discord-service/utils"
 	"github.com/bwmarrin/discordgo"
 	"github.com/sirupsen/logrus"
 )
 
-var NewDiscord = discordgo.New
-
 func SetupRegister() {
-	session, err := NewDiscord("Bot " + config.AppConfig.BOT_TOKEN)
+	session, err := utils.NewDiscordSession("Bot " + config.AppConfig.BOT_TOKEN)
```
Comment (iamitprakash) [inline commands/register/register.go:13]: same here, why there is change in implementation

File: commands/register/register_test.go [hunk 20]
```diff
@@ -3,25 +3,27 @@ package register
...
 	"github.com/stretchr/testify/assert"
 )
 
 func TestInit(t *testing.T) {
 
 	t.Run("should panic when SetupConnection returns an error", func(t *testing.T) {
-		originalNewDiscord := NewDiscord
-		defer func() { NewDiscord = originalNewDiscord }()
-		NewDiscord = func(token string) (s *discordgo.Session, err error) {
+		originalNewDiscord := utils.NewDiscordSession
```
Comment (iamitprakash) [inline commands/register/register_test.go:16]: same here

**Review by iamitprakash — CHANGES_REQUESTED** [review]

**Comment by iamitprakash** [comment]
I have done partial review, please address my comments and let me know for re-review

File: commands/handlers/listeningHandler_test.go (hunk 16 above)
Comment (joyguptaa) [reply commands/handlers/listeningHandler_test.go:82]: This comes from discord itself. There they are sending code as 0

File: commands/handlers/main_test.go (hunk 18 above)
Comment (joyguptaa) [reply commands/handlers/main_test.go:59]: In the current codebase, the retry mechanism is only used when we send data into the queue. It is not yet applied while performing the tasks.
Let me know if I should add it here or maybe at other places as well. I'll add it.

File: commands/main/register.go
Comment (joyguptaa) [reply commands/main/register.go:1]: This file was mistakenly duplicated. The current file, located inside commands/register/, performs the same task. It was needed earlier when we were using a separate command to register the commands. However, later on, based on @prakashchoudhary07 's suggestions, I moved that registration flow to the server start.

File: commands/register/register.go (hunk 19 above)
Comment (joyguptaa) [reply commands/register/register.go:13]: `NewDiscord` method was being used at multiple places earlier. So I've moved that into utils so that we do not have to create it at multiple places.

File: commands/register/register_test.go (hunk 20 above)
Comment (joyguptaa) [reply commands/register/register_test.go:16]: Here we are mocking the same util function which we have created to reduce repeated declarations.

File: commands/handlers/main.go (hunk 17 above)
Comment (joyguptaa) [reply commands/handlers/main.go:39]: We were creating it at multiple places, this exact same thing. So, for better reusability and a single source of truth, I've removed this from here and importing it from a common package.

File: commands/handlers/listeningHandler.go (hunk 15 above)
Comment (joyguptaa) [reply commands/handlers/listeningHandler.go:19]: Have updated the PR description as well as Issue Ticket
Ref : https://github.com/Real-Dev-Squad/discord-service/issues/72

File: commands/handlers/main.go (hunk 17 above)
Comment (iamitprakash) [reply commands/handlers/main.go:39]: this was test PR, changing the Title will not make useful, refactor needs to go in separate PR

File: commands/handlers/main_test.go (hunk 18 above)
Comment (iamitprakash) [reply commands/handlers/main_test.go:59]: and why ..?

File: commands/handlers/listeningHandler_test.go (hunk 16 above)
Comment (joyguptaa) [reply commands/handlers/listeningHandler_test.go:82]: Here we are making sure that we get to know the exact results why this particular task failed.

File: commands/handlers/main_test.go (hunk 18 above)
Comment (joyguptaa) [reply commands/handlers/main_test.go:59]: Must have been missed. Had only discussed retrying if sending data into the queue is failing. 
Let me know what you think and how we should cater this?

File: commands/handlers/listeningHandler_test.go (hunk 16 above)
Comment (devprakash0) [reply commands/handlers/listeningHandler_test.go:82]: 0 seems hardcoded doesn't make any point here, would you like me to share official documentation to prove your point

File: commands/handlers/main.go (hunk 17 above)
Comment (devprakash0) [reply commands/handlers/main.go:39]: no please go for two or more than PR 22 files will not get merged once, this is final
